# Code Refactoring and Shortening

## คำสั่งที่ควรรู้

## `lambda`

**Syntax**

`lambda arguments : expression`

`lambda` คือ Function รูปแบบย่อ ซึ่งจะรับ Arguments กี่ตัวก็ได้แต่สามารถส่งค่าออกได้แค่หนึ่งอย่างเท่านั้น

In [1]:
add = lambda a, b: a + b
add(3,4)

7

In [3]:
def add(a, b):
    """ Plus two numbers """
    return a + b
add(3,4)

7

จะเห็นได้ว่าการใช้ `lambda` จะใช้บรรทัดน้อยลงไปถึง 2 บรรทัดแต่ยังคงให้ผลลัพธ์เท่าเดิม เนื่องจากไม่จำเป็นต้องเขียน docstring และสามารถ Return ค่าได้ในบรรทัดเดียวกัน

---

## `if else`

**Syntax**

`statement if condition else statement` / `statement if condition else statement if statement else...`

ในส่วนนี้จะกล่าวถึงการใช้ `if else` ในรูปแบบย่อ ซึ่งสามารถใช้แทน `if else` ในรูปแบบปกติได้ในเกือบทุกกรณี

In [16]:
x = True
if x:
    print("Yes!")
else:
    print("Nah")

Yes!


In [17]:
x = True
print("Yes!" if x else "Nah")

Yes!


จะเห็นได้ว่าโค้ดทั้งสองให้ผลลัพธ์เหมือนกันแต่การใช้แบบย่อจะเขียนได้อยู่ในบรรทัดเดียว โดยสามารถมีกรณีอื่นๆเหมือนการใช้ `elif` ได้ตามปกติเหมือนดังโค้ดด้านล่าง

In [18]:
x = "B"
print("It's A" if x == "A" else "It's B" if x == "B" else "It's nor A or B")

It's B


---

## `for`

`for` ในที่นี้ไม่ได้กล่าวถึงการลูปแบบปกติ แต่จะเป็นการลูปเพื่อรับค่า, สร้าง list ฯลฯ โดยจะแสดงให้ให้ทั้งการใช้ `for` แบบปกติและแบบอยู่ใน list

In [12]:
x = []
for _ in range(0,3):
    x.append(input())
print(x)

1
2
3
['1', '2', '3']


คำสั่งด้านบนเป็นการใช้ `for` เพื่อรับค่า 3 ค่ามาเพิ่มไปใน list `x` ซึ่งสามาถเขียนแบบสั้นๆโดยนำ `[]` มาครอบได้แบบด้านล่างนี้

In [13]:
x = [input() for _ in range(0, 3)]
print(x)

Hello There!
General Kenobi!
You are a bold one.
['Hello There!', 'General Kenobi!', 'You are a bold one.']


การใช้ `for` ใน list เลยจะทำให้โด้ดสั้นลงได้และยังได้ผลลัพธ์เหมือนเดิม

In [14]:
x = [i for i in range(0, 5)]
print(x)

[0, 1, 2, 3, 4]


In [15]:
x = [chr(i) for i in range(65, 91)]
print(x)

['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']


และยังสามารถใช้เหมือนด้านบนเพื่อทำการสร้าง list ของค่าต่างๆได้

---

## `map()`

**Syntax**

`map(function, iterables)`

`map()` เป็น Built-in Function ชนิดหนึ่งที่ใช้ในการรัน Function ลงไปใน iterables หรือ Object ที่บรรจุค่าได้เช่น list และ tuple

**หมายเหตุ** - การจะแสดงผล map จำเป็นต้องแปลงให้เป็น list ก่อนโดยการครอบ `list()` ไว้ (ตัวอย่าง: `list(map(int, ["1", "2", "3"]))`)

In [5]:
def first(n):
    return n[0]
list(map(first, ["India", "Alpha", "Mike", "Tango", "Hotel", "Echo", "Sierra", "Echo", "November", "Alpha", "Tango", "Echo"]))

['I', 'A', 'M', 'T', 'H', 'E', 'S', 'E', 'N', 'A', 'T', 'E']

Function `first()` นั้นใช้เพื่อแปลง String ให้เหลือแค่ตัวอักษรตัวแรก ซึ่งเราก็สามารถใช้ `map()` ในการช่วงแปลง String ที่อยู่ใน list ให้เหลือต่ตัวแรกด้วย Function `first()` ของเราได้

In [9]:
x = [input(), input(), input()]
list(map(int, x))

44
122
120


[44, 122, 120]

จากโค้ดด้านบน list `x` เป็นการ input เข้ามา 3 ตัว ซึ่งค่า Default ของการ Input คือ String หากเราจะรับเป็น Int หรืออื่นๆก็จะต้องรับเป็น `int(input())` ซึ่งหากมี Input เป็นจำนวนมากและไม่สามารถใช้ `for` ได้การใช้ map แปลงค่าจะทำให้สั้นกว่า

---

## `join()`

**Syntax**

`string.join(iterable)`

`join()` เป็น Built-in Function ที่ใช้ในการรวม iterable เข้าด้วยกัน โดย string ด้านหน้าจะเป็นตัวกำหนดตัวที่ใช้เชื่อม

In [19]:
" ".join(["Not", "yet"])

'Not yet'

In [20]:
"+".join(["1","2","3"])

'1+2+3'

---

## `import`

**Syntax**

`from library import module`

การ import ด้วยรูปแบบนี้จะทำให้ไม่ต้อง import ทุกตัวของ Library นั้นมาและการเขียนจะสั้นลง

In [25]:
import math
math.pi

3.141592653589793

In [27]:
from math import pi
pi

3.141592653589793

จากตัวอย่างด้านบนหากเรามีการเรียกใช้งาน `pi` มากๆ แทนที่จะเขียนว่า `math.pi` จะเขียนแค่ `pi` ซึ่งจะทำให้โค้ดโดยรวมนั้นสั้นลง
ซึ่งการ import แบบนี้สามารถเลือกใช้หลายตัวได้ตามตัวอย่างด้านล่าง

In [29]:
from math import pi, sqrt
pi,sqrt(2)

(3.141592653589793, 1.4142135623730951)

---

## การใช้ `print()` ให้สั้น

หากค่ามีไม่มากให้รับค่าใน print เลย

In [35]:
print("Result: " + str(int(input()) + int(input())))

4
5
Result: 9


---
ใช้ `\n` ในการ print หลายบรรทัด แทนการใช้ print หลายบรรทัดแทน โดยจะใช้ร่วมกัน `sep` ในรูปแบบ `sep="\n"` ก็ได้

In [37]:
print("A\nB\nC")

A
B
C


In [30]:
print(1, 2, 3, sep="\n")

1
2
3


---
ใช้ `if else` ภายใน print

In [36]:
x = 1
print("Yes, it is" if x else "No, it isn't")

Yes, it is


---

## การนำ `lambda` `map()` และ `list()` มารวมกันเพื่อ `print()` ในบรรทัดเดียว

**Syntax**

`print(*list(map((lambda args: return), [value])))`

ตามรูปแบบดังกล่าวเราสามารถรับค่า > เข้าสูตร > print แสดงค่าได้ในบรรทัดเดียว

---
**หมายเหตุ** - หากมีการจำกัดห้ามใช้ `list()` หรือ `[]` ให้ใช้ Syntax ตามนี้แทน `print(*tuple(map((lambda args: return), (value, ))))`

---

**ตัวอย่าง 1:**

โดยด้านล่างจะเป็นโค้ดสำหรับรับคะแนนและบอกว่าผ่านหรือไม่ โดยจะผ่านเมื่อมากกว่า 60

จะเขียนเป็น 3 รูปแบบคือแบบปกติ, แบบการใช้ `lambda` ช่วย และแบบการใช้ `lambda` `map()` และ `list()` ร่วมกัน ตามลำดับ

In [42]:
def grade(x):
    """ Check your grade """
    if x >= 60:
        return "Pass"
    else:
        return "Fail"
    #ใช้เป็น return "Pass" if float(x) >= 60 else "Fail" ก็ได้
print(grade(float(input())))

75
Pass


In [39]:
grade = lambda x: "Pass" if float(x) >= 60 else "Fail"
print(grade(input()))

75
Pass


In [38]:
print(*list(map((lambda x: "Pass" if float(x) >= 60 else "Fail"), [input()])))

75
Pass


แต่นี้เป็นแค่ตัวอย่างเฉยๆ สำหรับโจทย์นี้หากจะให้สั้นที่สุดไม่จำเป็นต้องใช้ `lambda` ก็ได้

In [1]:
print("Pass" if float(input()) >= 60 else "Fail")

75
Pass


---
**ตัวอย่าง 2:**

โค้ดด้านล่างจะเป็นการรับคำมาสองคำและนำมาต่อกัน จากนั้นให้รับจำนวนเต็มมาเพื่อกำหนดว่าแสดงผลคำนั้นๆกี่รอบ โดยจะแสดงเป็น 3 รูปแบบเหมือนในตัวอย่าง 1

In [48]:
def repeat(i, j):
    """ Repeat i word j time(s)"""
    return (i + "\n") * j
print(repeat((input() + input()), int(input())))

Right Full Rudder!
Reverse Starboard Engine!
3
Right Full Rudder!Reverse Starboard Engine!
Right Full Rudder!Reverse Starboard Engine!
Right Full Rudder!Reverse Starboard Engine!



In [47]:
repeat = lambda i, j: (i + "\n") * j
print(repeat((input() + input()), int(input())))

Right Full Rudder!
Reverse Starboard Engine!
3
Right Full Rudder!Reverse Starboard Engine!
Right Full Rudder!Reverse Starboard Engine!
Right Full Rudder!Reverse Starboard Engine!



In [44]:
print(*list(map((lambda i, j: (i + "\n") * j), [input() + input()], [int(input())])))

Right Full Rudder!
Reverse Starboard Engine!
3
Right Full Rudder!Reverse Starboard Engine!
Right Full Rudder!Reverse Starboard Engine!
Right Full Rudder!Reverse Starboard Engine!



3 แบบด้านนี้ก็เป็นตัวอย่างเช่นเคย ซึ่งด้านบนก็สามารถเขียนให้สั้นกว่าโดยใช้แบบด้านล่างนี้ก็ได้

In [1]:
print((input() + input() + "\n") * int(input()))

Right Full Rudder!
Reverse Starboard Engine!
3
Right Full Rudder!Reverse Starboard Engine!
Right Full Rudder!Reverse Starboard Engine!
Right Full Rudder!Reverse Starboard Engine!



---
**ตัวอย่าง 3:**

โด้ดในข้อนี้จะเป็นการรับจำนวนเต็มมา 5 ค่า และให้แสดงค่าที่เป็นจำนวนคี่ซึ่งจะมีแค่จำนวนเดียวจาก 5 จำนวน

In [4]:
print("Odd is: " + str(sum(list(map(lambda a: a if a%2 else 0, list(map(int, ([input(), input(), input(), input(), input()]))))))))

12
30
8
67
102
Odd is: 67


โค้ดด้านบนนั้นทำงานอย่างไร ให้ดูใน Arguments ที่สองของ map กันก่อน (มีการตัด `"Odd is: " + str()` ออก)

`print(sum(list(map(lambda a: a if a%2 else 0,`**`list(map(int, ([input(), input(), input(), input(), input()])))`**`))))`

ตรงนี้คือส่วนที่เรารับค่าทั้ง 5 ตัวเข้ามา(โดยไม่ใช้ `for`) ซึ่งรับมาเป็น `string` เลยนำ `map(int, iter)` มาครอบเพื่อแปลงค่าข้างในให้เป็น `int` ให้หมด และครอบด้วย `list()` อีกที่เพื่อให้ map object กลายเป็น list

ส่วนต่อไปคือ lambdaซึ่งเป็น Arguments ตัวแรกของ map

`print(sum(list(map(`**`lambda a: a if a%2 else 0`**`, list(map(int, ([input(), input(), input(), input(), input()])))))))`

การใช้ `map()` จะเป็นการนำ function เข้าไปรันกับค่าภายใน list ทีละตัว ซึ่ง `lambda` ก็คือ function ชนิดหนึ่ง โดย lambda ด้านบนของเรานั้นจะเป็นการรับค่า `a` เข้ามา ซึ่งก็คือค่าต่างๆใน list และใส่ `if else` ว่า **"ให้คืนค่าเป็น a ถ้า a%2 = 1 หากเป็นแบบอื่นๆให้คืนค่าเป็น 0"** ดังนั้นค่าที่เข้ามาหากค่านั้นเป็นจำนวนคู่ค่านั้นก็จะกลายเป็น 0 แต่ถ้าเป็นคี่ค่าก็จะไม่เปลี่ยน

ทำให้จากที่เรา input ไปด้านบนคือ `[12, 30, 8, 67, 102]` จะกลายเป็น `[0, 0, 0, 67, 0]`

จากนั้นครอบ `map()` ด้วย `list()` เพื่อให้กลายเป็น list จากนั้น `sum()` เพื่อรวมค่าใน list เข้าด้วยกันและ `print()` ค่านั้นออกมาโดยค่านั้นก็คือ `67`